### PR TITLE
Linux: fall back to cpuinfo on slow scaling_cur_freq read

### PR DESCRIPTION
**Pull request by cgzones**
On some AMD CPUs read()ing scaling_cur_freq is quite slow (> 1ms).
This delay accumulates for every cpu.
If the read on CPU 0 takes longer than 500us bail out and fall back to
reading the frequencies from /proc/cpuinfo.
Once the condition has been met, bail out early for the next couple of
scans.

Closes: #471

### CODE REVIEW

File: linux/LinuxProcessList.c [hunk 1]
```diff
@@ -1727,10 +1728,28 @@ static int scanCPUFreqencyFromSysCPUFreq(LinuxProcessList* this) {
    int numCPUsWithFrequency = 0;
    unsigned long totalFrequency = 0;
 
+   /*
+    * On some AMD CPUs read()ing scaling_cur_freq is quite slow (> 1ms).
```
Comment (KonstantinPlotnikov) [inline linux/LinuxProcessList.c:1732]: Just to be accurate.
As you could see in https://github.com/htop-dev/htop/issues/471#issuecomment-759378878 this is not AMD specific issue. At least some of the Intel Xeon CPUs also share such behavior.

**Comment by KonstantinPlotnikov** [comment]
This fix completely solves issue #471 for me. Thank you!

**Review by BenBE — COMMENTED** [review]
LGTM. Maybe update the comment to include this seems to be the case whenever a non-`intel_pstate` driver for CPU frequency is used (affects AMD and some Intel [Xeon] CPUs).

**Comment by fasterit** [comment]
@BenBE: intel_pstate was also observed slow in #471 (Intel Core i7-7700 case).

I'd still be interested _why_ we see the difference between 15ms and ~0ms. It does not seem to be dependent on the driver. Is it consistent for each system?
May be we are just seeing the kernel context switching and scheduling another application while we are suspended for the `read` syscall. In that case testing the first read may not be much better than guessing.

Related:
`readfile()` will probably be useful in the future cf. `https://git.kernel.org/pub/scm/linux/kernel/git/gregkh/driver-core.git/commit/?h=readfile&id=3317335407261fd5f525bd5fd63ade139fed6ccb`

**Comment by BenBE** [comment]
@fasterit:
> Related:
> `readfile()` will probably be useful in the future cf. `https://git.kernel.org/pub/scm/linux/kernel/git/gregkh/driver-core.git/commit/?h=readfile&id=3317335407261fd5f525bd5fd63ade139fed6ccb`

Which kernel release does this correspond to? Looks like it might be worth to check for this syscall in `xReadfile`.

**Comment by fasterit** [comment]
@BenBE: None yet, Linus still needs to merge that one. Greg has had it out since ~07/2020 and it seems rather ready, so may be as early as  5.12.

**Comment by clain-kp** [comment]
> I'd still be interested _why_ we see the difference between 15ms and ~0ms. It does not seem to be dependent on the driver. Is it consistent for each system?
> May be we are just seeing the kernel context switching and scheduling another application while we are suspended for the read syscall. In that case testing the first read may not be much better than guessing.

I have 11 AMD EPYC ( 1 Zen + 10 Zen2) systems each one of those has this issue. And those systems are idle most of the time. Reboot does not affect this.
I did not run specific automatic test to collect statistics on single machine, but every manual check gives pretty consistent result ( 0 of ~15 ms).

As for the reason of this: I do not know where to look. If you have any idea I could collect almost any data from "slow" CPUs.

**Comment by KonstantinPlotnikov** [comment]
> I'd still be interested why we see the difference between 15ms and ~0ms. It does not seem to be dependent on the driver. Is it consistent for each system?
> May be we are just seeing the kernel context switching and scheduling another application while we are suspended for the read syscall. In that case testing the first read may not be much better than guessing.

I have 11 AMD EPYC ( 1 Zen + 10 Zen2) systems each one of those has this issue. And those systems are idle most of the time. Reboot does not affect this.
I did not run specific automatic test to collect statistics on single machine, but every manual check gives pretty consistent result ( 0 of ~15 ms).

As for the reason of this: I do not know where to look. If you have any idea I could collect almost any data from "slow" CPUs.

**Comment by cgzones** [comment]
Once the `readfile(2)` syscall gets merged I am looking forward to use it in our `xReadfile()` wrapper conditionally (and maybe convert some frequently used `fopen()` + `fread()/fscanf()` + `fclose()` pairs to it).
But I fear that has nothing to do with this slowdown, as the data shows the `open` and `close` calls on the `scaling_cur_freq` inodes are fast; it's only the `read` call that's slow, probably because the driver takes that time to gather the data from the actual hardware (maybe the frequency is not stored/cached and can only be retrieved after a context switch).

**Comment by BenBE** [comment]
> As for the reason of this: I do not know where to look. If you have any idea I could collect almost any data from "slow" CPUs.

I think the most interesting data here are likely flame graphs[1] of the kernel stack while executing this particular read syscall. To make that more obvious you'd most likely would write a small program that just polls that value in a loop and measure it with `perf` or some kernel tracing to get the flame graph data[2]. While this will likely give results pointing to the culprit, it's unlikely we can do anything about it in userspace; thus the results are likely to be reported to The Kernel People™ …

[1] http://www.brendangregg.com/flamegraphs.html
[2] https://www.gabriel.urdhr.fr/2014/05/23/flamegraph/#stacksampling

**Pull request merged**